src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.DecisionDagRewriter.cs

```diff
@@ -865,59 +870,231 @@ private void EnsureStringHashFunction(int labelsCount, SyntaxNode syntaxNode)
+                    var found = whenNodeMap.TryGetValue(whenNode, out var whenNodeInfo);
+                    Debug.Assert(found);
+
+                    // whenNodeIdentifierLocal = whenNodeIdentifier;
+                    sectionBuilder.Add(_factory.Assignment(_factory.Local(_whenNodeIdentifierLocal), _factory.Literal(whenNodeInfo.WhenNodeIdentifier)));
+
+                    // goto labelToWhenExpression;
+                    sectionBuilder.Add(_factory.Goto(whenNodeInfo.LabelToWhenExpression));
+
+                    return true;
+                }
+
+                void lowerWhenExpressionIfShared(BoundExpression whenExpression, LabelSymbol labelToWhenExpression, List<BoundWhenDecisionDagNode> whenNodes)
                 {
-                    _factory.Syntax = whenClause.Syntax;
-                    BoundStatement conditionalGoto = _factory.ConditionalGoto(_localRewriter.VisitExpression(whenClause.WhenExpression), trueLabel, jumpIfTrue: true);
+                    if (!isSharedWhenExpression(whenExpression))
+                    {
+                        return;
+                    }
+
+                    var whenClauseSyntax = whenNodes[0].Syntax;
+                    var whenTrueLabel = GetDagNodeLabel(whenNodes[0].WhenTrue);
+                    Debug.Assert(whenNodes.All(n => n.Syntax == whenClauseSyntax));
+                    Debug.Assert(whenNodes.All(n => n.WhenExpression == whenExpression));
+                    Debug.Assert(whenNodes.All(n => n.Bindings == whenNodes[0].Bindings));
+                    Debug.Assert(whenNodes.All(n => whenTrueLabel == GetDagNodeLabel(whenNodes[0].WhenTrue)));
+
+                    ArrayBuilder<BoundStatement> sectionBuilder = BuilderForSection(whenClauseSyntax);
+                    sectionBuilder.Add(_factory.Label(labelToWhenExpression));
+                    lowerBindings(whenNodes[0].Bindings, sectionBuilder);
+
+                    // if (loweredWhenExpression)
+                    // {
+                    //   jump to whenTrue label
+                    // }
+                    _factory.Syntax = whenClauseSyntax;
+                    BoundStatement conditionalGoto = _factory.ConditionalGoto(_localRewriter.VisitExpression(whenExpression), whenTrueLabel, jumpIfTrue: true);
 
                     // Only add instrumentation (such as a sequence point) if the node is not compiler-generated.
-                    if (GenerateInstrumentation && !whenClause.WhenExpression.WasCompilerGenerated)
+                    if (GenerateInstrumentation && !whenExpression.WasCompilerGenerated)
                     {
-                        conditionalGoto = _localRewriter._instrumenter.InstrumentSwitchWhenClauseConditionalGotoBody(whenClause.WhenExpression, conditionalGoto);
+                        conditionalGoto = _localRewriter._instrumenter.InstrumentSwitchWhenClauseConditionalGotoBody(whenExpression, conditionalGoto);
                     }
 
                     sectionBuilder.Add(conditionalGoto);
```

Comment:
Consider extracting to a local function, for use here and in `lowerWhenClause()`.

Reply:
I didn't consider it and was on the fence... Let's see how it looks